tantar/src/main/java/org/opencadc/tantar/BucketValidator.java

```diff
@@ -588,82 +591,139 @@ public void updateArtifact(final Artifact artifact, final StorageMetadata storag
-            Iterator<StorageMetadata> storageMetadataIterator =
-                    Subject.doAs(runUser, (PrivilegedExceptionAction<Iterator<StorageMetadata>>) () ->
-                                                      storageAdapter.iterator(bucketPrefixIterator.next()));
+            ResourceIterator<Artifact> artifactIterator = iteratorDAO.storedIterator(bucketPrefixIterator.next());
 
             @Override
             public boolean hasNext() {
-                if (storageMetadataIterator.hasNext()) {
+                if (artifactIterator.hasNext()) {
                     return true;
                 } else if (bucketPrefixIterator.hasNext()) {
-                    try {
-                        storageMetadataIterator =
-                                Subject.doAs(runUser, (PrivilegedExceptionAction<Iterator<StorageMetadata>>) () ->
-                                                                  storageAdapter.iterator(bucketPrefixIterator.next()));
-                    } catch (PrivilegedActionException exception) {
-                        throw new IllegalStateException(exception);
-                    }
+                    artifactIterator = iteratorDAO.storedIterator(bucketPrefixIterator.next());
                     return hasNext();
                 } else {
                     return false;
                 }
             }
 
             @Override
-            public StorageMetadata next() {
-                return storageMetadataIterator.next();
+            public Artifact next() {
+                return artifactIterator.next();
             }
         };
     }
 
     /**
-     * Iterate over the Artifact instances from the Storage Inventory database.  The consumer of this Iterator will
-     * assume that the StorageLocation is never null in the Artifact.
-     * <p/>
-     * Tests can override this for convenience.
+     * Check the StorageMetadata for a matching ObsoleteStorageLocation and if found
+     * delete the file from storage and delete the ObsoleteStorageLocation.
+     */
+    boolean isObsoleteStorageLocation(StorageMetadata storageMetadata) {
+        ObsoleteStorageLocation obsoleteStorageLocation;
+        try {
+            obsoleteStorageLocation = this.obsoleteStorageLocationDAO.get(storageMetadata.getStorageLocation());
+        } catch (Exception e) {
+            throw new IllegalStateException(
+                String.format("query failed for ObsoleteStorageLocation %s for file %s",
+                              storageMetadata.getStorageLocation().getStorageID().toASCIIString(),
+                              storageMetadata.artifactURI.toASCIIString()), e);
+        }
+        if (obsoleteStorageLocation != null && canTakeAction()) {
+            try {
+                delete(storageMetadata);
+                this.obsoleteStorageLocationDAO.delete(obsoleteStorageLocation.getID());
+                LOGGER.info(String.format("deleted obsolete file %s at %s",
```

Comment:
The log message is inside the canTakeAction() so nothing is reported in reportOnly mode; that doesn't seem useful. An operator will run this with reportOnly=true, see nothing in the output, then run it for real and possibly see messages about files being deleted. Surprise!!

I don't know about the other uses of canTakeAction() vs logging, but if they are like this they suck. In general, you want the info log message before the attempted action, so "delete obsolete file ..." (not past tense). Further, since this message says it is deleting artifactURI it is very misleading: **artifactURI is not obsolete** and probably still exists (because an OSL is created when an artifact is overwritten as well as deleted). So something like this would be correct:
```
delete obsolete object: {storageLocation} (was: {artifactURI})
```
And you need to log the whole StorageLocation, not just the storageID - yeah, the default toString() is OK for now since it isn't obvious how to write storageBucket and storageID clearly without an unambiguous separator.

Reply:
Most tantor methods that do work start with an `if (canTakeAction)` and there is no logging outside of that `if` so yeah running in dryrun mode is not going to tell the operator much.

Reply:
Updated the logging to always log the intent in isObsoleteStorageLocation().